src/vm/eventpipeprovider.cpp

```diff
@@ -45,8 +45,11 @@ EventPipeProvider::~EventPipeProvider()
     // Free all of the events.
     if(m_pEventList != NULL)
     {
-        // Take the lock before manipulating the list.
-        CrstHolder _crst(EventPipe::GetLock());
+        // Here we manipulate the list without taking a lock
+        // This is because the lock is HOST_BREAKABLE and may throw
```

Comment:
Same thing here.  This is not safe to do without taking the lock because it's possible that a new event is being added during this operation, which would be unsafe without holding the lock.  I think you can use the same pattern here - wrap in a try/catch and leak the memory if the lock is broken by the host.

Reply:
I tried taking a lock within a EX_TRY block here. I got this assert

```
Assert failure(PID 5536 [0x000015a0], Thread: 29512 [0x7348]): CONTRACT VIOLATION by CrstBase::Enter at "c:\users\vigraf\coreclr\src\vm\crst.cpp" @ 251

You cannot enter a lock in a GC_NOTRIGGER + MODE_COOPERATIVE region.

VIOLATED-->  CONTRACT in EventPipeProvider::~EventPipeProvider at "c:\users\vigraf\coreclr\src\vm\eventpipeprovider.cpp" @ 44
                        CONTRACT in EventPipeConfiguration::~EventPipeConfiguration at "c:\users\vigraf\coreclr\src\vm\eventpipeconfiguration.cpp" @ 35
                        CONTRACT in EventPipe::Shutdown at "c:\users\vigraf\coreclr\src\vm\eventpipe.cpp" @ 229
                        CONTRACT in EEShutDownHelper at "c:\users\vigraf\coreclr\src\vm\ceemain.cpp" @ 1554
                        GCX_COOP in EEShutDown at "c:\users\vigraf\coreclr\src\vm\ceemain.cpp" @ 2218
                        CONTRACT in EEShutDown at "c:\users\vigraf\coreclr\src\vm\ceemain.cpp" @ 2136



CORECLR! CONTRACT_ASSERT + 0x34A (0x00007ffb`92bdaada)
CORECLR! CrstBase::Enter + 0x15E (0x00007ffb`92d0fbce)
CORECLR! CrstBase::AcquireLock + 0x15 (0x00007ffb`92c87405)
CORECLR! CrstBase::CrstHolder::CrstHolder + 0x25 (0x00007ffb`92c86175)
CORECLR! EventPipeProvider::~EventPipeProvider + 0x1B8 (0x00007ffb`93581a18)
CORECLR! EventPipeProvider::`scalar deleting destructor' + 0x17 (0x00007ffb`9357eaa7)
CORECLR! EventPipeConfiguration::~EventPipeConfiguration + 0x188 (0x00007ffb`9357e3b8)
CORECLR! EventPipeConfiguration::`scalar deleting destructor' + 0x17 (0x00007ffb`93158627)
CORECLR! EventPipe::Shutdown + 0x1F0 (0x00007ffb`9315ab00)
CORECLR! EEShutDownHelper + 0x18C (0x00007ffb`92c9441c)
    File: c:\users\vigraf\coreclr\src\vm\crst.cpp Line: 251
    Image: C:\Users\vigraf\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\CoreRun.exe
```

Do you have an idea for what to do from here?

Reply:
Can you mark ~EventPipeProvider as GC_TRIGGERS or is that not OK elsewhere?  I think that would do it (or atleast get us started).